PR: https://github.com/odoo/odoo/pull/236080

From: 2fd2df6b997fc860275ffc95a1c4f8683fcc70d4
From: alan-odoo
Date: 2026-09-02 08:30:17

Breaking data model changes scores: del:2 + add:0, change matches:
-    invoiced_amount = fields.Integer(
-    company_id = fields.Many2one(

Total Changes: 42

[FIX] *: compute invoiced amounts using exchange rates

* = account, mass_mailing_sale, sale

This commit fixes the total invoiced amounts displayed in the stat buttons of
the mailing, campaign and partner forms. Previously, the currency of the
invoices was not taken into account. Then, if the invoices involve multiple
currencies, the amounts were add up without using currency rates making the
total meaningless. To prevent that, all invoiced amounts are now converted
into a unique currency to calculate the total.

The unique currency is the one of the environment's company for the campaigns
and the mailings, as they have no company. For the partners, the currency of
their company is used in priority if they have one.

Previously, UTM campaigns were linked to a company. This relationship was done
only to display the currency of their invoiced amounts. However, campaigns
should not have a company as every users whatever the company are able to read
and modify them. Therefore, to display a total invoiced amount that users can
easily understand, the currency must not be the one of the company which
created the campaign, but the one of the environment's company. For this
reason, UtmCampaing.currency_id is now computed depending on the environment's
company and is no longer defined with the company_id field. Since this last
field is not used anymore by the currency_id field, it has no utility anymore
and has been removed.

To fix the invoiced amounts, SQL have been preferred to Python for
optimization reasons as currency conversions involve several tables. It is
particularly important for res.partner model as the unique currency varies for
partners as each one may belongs to a different company, this implies extra
steps not requested by mailing.mailing and utm.campaign which only need the
environment's company.
Here is a quick and general description of the SQL queries. First, they get
the currency rates of the billing company with the currency of the
environment's company or the partner's company depending on the context. If
several currency rates with the same company and currency are present, only
the last one created before the invoice is kept, as it is the one active at
that time. And finally, the sum is done after the conversions' executions.

Also, the currency rates can be modified by users. This means that the
converted amount at the time of the invoice creation may have different than
the one calculated by the compute methods. However, this should not happen
frequently and those total invoiced amounts are only displayed as a preview
but not needed to calculate data.

To reproduce:
Two companies with different currencies must be created. The user for the
testing must be linked to both of them and they must be activated with the
select in the upper right corner of page.

Create a partner, a campaign and a mailing and linked them to two invoices
using the currencies of the two companies. On their forms, the amount is the
sum of the prices without any conversion.

Enterprise PR: https://github.com/odoo/enterprise/pull/99697

Task-5258611

Part-of: odoo/odoo#236080
Related: odoo/enterprise#99697
Related: odoo/upgrade#8989
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/sale/models/utm_campaign.py
+++ b/addons/sale/models/utm_campaign.py
@@ -1,6 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import api, fields, models
 from odoo.tools import SQL
 
 
@@ -13,22 +13,28 @@ class UtmCampaign(models.Model):
         compute_sudo=True,
         groups="sales_team.group_sale_salesman",
     )
-    invoiced_amount = fields.Integer(
+    invoiced_amount = fields.Monetary(
         string="Revenues generated by the campaign",
         compute="_compute_sale_invoiced_amount",
         compute_sudo=True,
         groups="sales_team.group_sale_salesman",
-    )
-    company_id = fields.Many2one(
-        "res.company",
-        string="Company",
-        index="btree_not_null",
-        default=lambda self: self.env.company,
+        currency_field='currency_id',
     )
     currency_id = fields.Many2one(
-        "res.currency", related="company_id.currency_id", string="Currency"
+        "res.currency",
+        string="Currency",
+        compute="_compute_currency_id",
+        compute_sql="_compute_sql_currency_id",
+        compute_sudo=True,
     )
 
+    @api.depends_context('company')
+    def _compute_currency_id(self):
+        self.currency_id = self.env.company.currency_id
+
+    def _compute_sql_currency_id(self, table):
+        return SQL("%s", self.env.company.currency_id.id)
+
     def _compute_quotation_count(self):
         quotation_data = self.env["sale.order"]._read_group(
             [("campaign_id", "in", self.ids)], ["campaign_id"], ["__count"]
@@ -49,17 +55,29 @@ class UtmCampaign(models.Model):
             query_res = self.env.execute_query_dict(
                 SQL(
                     """
-                    SELECT move.campaign_id, -SUM(line.balance) as price_subtotal
+                    SELECT move.campaign_id, -SUM(line.balance * COALESCE(currency_rate.rate, 1)) as price_subtotal
                       FROM account_move_line line
                 INNER JOIN account_move move ON line.move_id = move.id
+                      /* To use the exchange rate effective at the creation of the invoice. */
+                      LEFT JOIN LATERAL (
+                          SELECT rate
+                            FROM res_currency_rate
+                           WHERE company_id = line.company_id
+                             AND currency_id = %(currency_id)s
+                             AND name < line.date
+                           ORDER BY name DESC
+                           LIMIT 1
+                      ) currency_rate
+                        ON TRUE
                      WHERE move.state not in ('draft', 'cancel')
-                       AND move.campaign_id IN %s
+                       AND move.campaign_id IN %(campaign_ids)s
                        AND move.move_type IN ('out_invoice', 'out_refund', 'in_invoice', 'in_refund', 'out_receipt', 'in_receipt')
                        AND line.account_id IS NOT NULL
                        AND line.display_type = 'product'
                   GROUP BY move.campaign_id
                     """,  # noqa: E501
-                    tuple(self.ids),
+                    campaign_ids=tuple(self.ids),
+                    currency_id=self.env.company.currency_id.id,
                 )
             )
         else:
